File: app/src/main/java/com/sardox/timestamper/recyclerview/CategoryAdapter.java
package com.sardox.timestamper.recyclerview;

import android.content.Context;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.sardox.timestamper.R;
import com.sardox.timestamper.objects.Category;
import com.sardox.timestamper.utils.Consumer;

import java.util.List;

public class CategoryAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final int CATEGORY_ITEM = 0;
    private final int ADD_BUTTON_ITEM = 1;
    private final int categoryColorSelected;
    private final int categoryColorDeselected;

    private Category selectedCategory = Category.Default;
    private Consumer<Category> categoryChangedCallback;
    private Consumer<Void> categoryAddCallback;
    private List<Category> categories;

    public CategoryAdapter(List<Category> categories, Consumer<Category> categoryChangedCallback, Consumer<Void> categoryAddCallback, Context context) {
        this.categoryAddCallback = categoryAddCallback;
        this.categoryChangedCallback = categoryChangedCallback;
        this.categories = categories;

        categoryColorSelected = ContextCompat.getColor(context, R.color.category_selected);
        categoryColorDeselected = ContextCompat.getColor(context, R.color.colorPrimary);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        switch (viewType) {
            case CATEGORY_ITEM: {
                View view = LayoutInflater.from(viewGroup.getContext())
                        .inflate(R.layout.recyclerview_category, viewGroup,
                                false);
                return new CategoryViewHolder(view);
            }
            case ADD_BUTTON_ITEM: {
                View view = LayoutInflater.from(viewGroup.getContext())
                        .inflate(R.layout.recyclerview_category_add_button, viewGroup,
                                false);
                return new AddCategoryButtonViewHolder(view);
            }
        }
        return null;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {

        switch (holder.getItemViewType()) {
            case CATEGORY_ITEM:
                Category category = categories.get(position);
                CategoryViewHolder categoryViewHolder = (CategoryViewHolder) holder;
                categoryViewHolder.mCategoryTextView.setText(category.getName());
                if (selectedCategory.getCategoryID().equals(category.getCategoryID())){
                    categoryViewHolder.category_underline.setBackgroundColor(categoryColorSelected);
                } else {
                    categoryViewHolder.category_underline.setBackgroundColor(categoryColorDeselected);
                }
                break;
            case ADD_BUTTON_ITEM:
                break;
        }
    }


    @Override
    public int getItemViewType(int position) {
        if (categories.size() == position) {
            return ADD_BUTTON_ITEM;
        } else {
            return CATEGORY_ITEM;
        }
    }

    public void setSelectedCategory(Category selected_category) {
        this.selectedCategory = selected_category;
    }

    @Override
    public int getItemCount() {
        return categories.size()+1;
    }


    class CategoryViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private TextView mCategoryTextView;
        private LinearLayout category_underline;

        CategoryViewHolder(View itemView) {
            super(itemView);
            mCategoryTextView = itemView.findViewById(R.id.CategoryTextView);
            category_underline = itemView.findViewById(R.id.recycler_category_underline);
            mCategoryTextView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            selectedCategory = categories.get(getAdapterPosition());
            categoryChangedCallback.accept(selectedCategory);
            notifyDataSetChanged();
        }
    }

    class AddCategoryButtonViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private TextView mCategoryAdd;

        AddCategoryButtonViewHolder(View itemView) {
            super(itemView);
            mCategoryAdd = itemView.findViewById(R.id.category_item_button);
            mCategoryAdd.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            categoryAddCallback.accept(null);
        }
    }
}

File: app/src/main/java/com/sardox/timestamper/recyclerview/CategoryAdapter.kt
package com.sardox.timestamper.recyclerview

import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

import com.sardox.timestamper.R
import com.sardox.timestamper.objects.Category
import com.sardox.timestamper.utils.CategoryUpdatesInterface
import com.sardox.timestamper.utils.TimestampIcon

class CategoryAdapter(private val categories: List<Category>, private val categoryChangedCallback: CategoryUpdatesInterface, private val icons: List<TimestampIcon>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val CATEGORY_ITEM = 0
    private val ADD_BUTTON_ITEM = 1

    private var selectedCategory = Category.Default

    fun selectedCategoryPosition(): Int {
        val pos = categories.indexOf(selectedCategory)
        return if (pos == -1)
            0
        else
            pos
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            CATEGORY_ITEM -> {
                CategoryViewHolder(LayoutInflater.from(viewGroup.context)
                        .inflate(R.layout.recyclerview_category, viewGroup,
                                false))
            }
            else -> {
                AddCategoryButtonViewHolder(LayoutInflater.from(viewGroup.context)
                        .inflate(R.layout.recyclerview_category_add_button, viewGroup,
                                false))
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder.itemViewType) {
            CATEGORY_ITEM -> {
                val category = categories[position]
                val categoryViewHolder = holder as CategoryViewHolder
                categoryViewHolder.mCategoryTextView.text = category.name
                categoryViewHolder.categoryIcon.setImageDrawable(ContextCompat.getDrawable(holder.categoryIcon.context, icons[category.icon_id].drawable_id))
                categoryViewHolder.categoryContainer.isSelected = selectedCategory.categoryID == category.categoryID

                if (selectedCategory.categoryID == category.categoryID) {
                    categoryViewHolder.categoryContainer.isSelected = true
                    categoryViewHolder.categoryRemove.visibility = View.VISIBLE
                } else {
                    categoryViewHolder.categoryContainer.isSelected = false
                    categoryViewHolder.categoryRemove.visibility = View.GONE
                }
            }
            ADD_BUTTON_ITEM -> {
            }
        }
    }


    override fun getItemViewType(position: Int): Int {
        return if (categories.size == position) {
            ADD_BUTTON_ITEM
        } else {
            CATEGORY_ITEM
        }
    }

    fun setSelectedCategory(selectedCategory: Category) {
        this.selectedCategory = selectedCategory
    }

    override fun getItemCount(): Int {
        return categories.size + 1
    }

    internal inner class CategoryViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        var mCategoryTextView: TextView = itemView.findViewById(R.id.categoryNameTextView)
        var categoryIcon: ImageView = itemView.findViewById(R.id.categoryImage)
        var categoryRemove: ImageView = itemView.findViewById(R.id.recycler_menu_remove)
        var categoryContainer: LinearLayout = itemView.findViewById(R.id.categoryContainer)

        init {
            mCategoryTextView.setOnClickListener(this)
            categoryRemove.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            when (view) {
                mCategoryTextView -> {
                    selectedCategory = categories[adapterPosition]
                    categoryChangedCallback.onCategoryChanged(selectedCategory)
                    notifyDataSetChanged()
                }
                categoryRemove -> categoryChangedCallback.onCategoryRemove(selectedCategory)
            }
        }
    }

    internal inner class AddCategoryButtonViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        private val mCategoryAdd: TextView = itemView.findViewById(R.id.category_item_button)

        init {
            mCategoryAdd.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            categoryChangedCallback.onCategoryAdded()
        }
    }
}